helper/storagepacker/storagepacker_v2.go

```diff
@@ -0,0 +1,765 @@
+			}
+		}()
+	}
+
+	numShards := int(math.Pow(2.0, float64(s.BucketShardBits)))
+
+	// Create the shards and lock them
+	s.Logger.Info("sharding bucket", "bucket_key", bucket.Key, "num_shards", numShards)
+	defer s.Logger.Info("sharding bucket process exited", "bucket_key", bucket.Key)
+
+	shardLocks[bucket.LockEntry] = struct{}{}
+
+	shards := make(map[string]*LockedBucket, numShards)
+	for i := 0; i < numShards; i++ {
+		shardKey := fmt.Sprintf("%x", i)
+		lock := locksutil.LockForKey(s.storageLocks, cacheKey+shardKey)
+		shardedBucket := &LockedBucket{
+			LockEntry: lock,
+			Bucket: &Bucket{
+				Key:     fmt.Sprintf("%s/%s", bucket.Key, shardKey),
+				ItemMap: make(map[string]*any.Any),
+			},
+		}
+		shards[shardKey] = shardedBucket
+		// If it was equal we'd be locked already
+		s.Logger.Trace("created shard", "shard_key", shardKey)
+		// Don't try to lock the same lock twice in case it hashes that way
+		if _, ok := shardLocks[lock]; !ok {
+			s.Logger.Trace("locking lock", "shard_key", shardKey)
+			lock.Lock()
+			shardLocks[lock] = struct{}{}
+		}
+	}
+
+	s.Logger.Debug("resilvering items")
+
+	parentPrefix := s.GetCacheKey(bucket.Key)
+	// Resilver the items
+	for k, v := range bucket.ItemMap {
+		itemKey := strings.TrimPrefix(k, parentPrefix)[0 : s.BucketShardBits/4]
+		s.Logger.Trace("resilvering item", "parent_prefix", parentPrefix, "item_id", k, "item_key", itemKey)
+		// Sanity check
+		childBucket, ok := shards[itemKey]
+		if !ok {
+			// We didn't complete sharding so don't make other parts of the
+			// code think that it completed
+			s.Logger.Error("failed to find sharded storagepacker bucket", "bucket_key", bucket.Key, "item_key", itemKey)
+			return errors.New("failed to shard storagepacker bucket")
+		}
+		childBucket.ItemMap[k] = v
+	}
+
+	s.Logger.Debug("storing sharded buckets")
+
+	// Ensure we can write all of these buckets. Create a cleanup function if not.
+	retErr := new(multierror.Error)
+	cleanupStorage := func() {
+		for _, v := range shards {
+			if err := s.BucketStorageView.Delete(ctx, v.Key); err != nil {
+				retErr = multierror.Append(retErr, err)
+				// Don't exit out, clean up as much as possible
+			}
+		}
+	}
+	for k, v := range shards {
+		s.Logger.Trace("storing bucket", "shard", k)
+		if err := s.storeBucket(ctx, v, false); err != nil {
+			s.Logger.Debug("encountered error", "shard", k)
+			retErr = multierror.Append(retErr, err)
+			cleanupStorage()
+			return retErr
+		}
+	}
+
+	cleanupCache := func() {
+		for _, v := range shards {
+			s.bucketsCache.Delete(s.GetCacheKey(v.Key))
+		}
+	}
+	// Add to the cache. It's not too late to back out, via the cleanup cache
+	// function. We hold the lock while storing the updated original bucket so
+	// that nobody accesses it in an inconsistent state.
+	s.Logger.Debug("updating cache")
+	s.bucketsCacheLock.Lock()
+	{
+		for _, v := range shards {
+			s.bucketsCache.Insert(s.GetCacheKey(v.Key), v)
```

Comment:
I noticed two things which possibly needs some updating. 
1) PutBucket, after calling storeBucket inserts the parent bucket into the cache. This is fine if the sharding doesn't happen. If the bucket gets sharded, it should not be added into the cache after a successful storeBucket.
2) After adding all the sharded buckets into the cache, the parent bucket should be removed from the cache. No?